src/ngraph/runtime/gpu/nvrtc/helpers.hpp

```diff
@@ -0,0 +1,208 @@
+        "mov.b64 {lo32, hi32}, res64;\n\t"
+        "selp.u32 hi32, hi32, %1, p;\n\t"
+        "shr.u32 %0, hi32, %3;\n\t"
+        "}" : "=r"(result) : "r"(value), "r"(magic), "r"(shift));
+    return result;
+}
+
+__device__ __forceinline__ void idiv_fast(int numerator, int denominator, float rcp,
+                                          int& result, int& remainder)
+{
+    result = (int)((float)numerator * rcp);
+    remainder = numerator - (result * denominator);
+    result = (remainder >= denominator) ? (result + 1) : result;
+    remainder = (remainder >= denominator) ? (remainder - denominator) : remainder;
+}
+
+__device__ __forceinline__ int mod16(int numerator, int div, int maxdiv)
+{
+    int res;
+    asm("vmad.s32.u32.u32 %0, -%1.h0, %2.h0, %3;" : "=r"(res) : "r"(div), "r"(maxdiv), "r"(numerator));
+    return res;
+}
+__device__ __forceinline__ int mad16(int a, int b, int c)
+{
+    int res;
+    asm("vmad.s32.u32.u32 %0, %1.h0, %2.h0, %3;" : "=r"(res) : "r"(a), "r"(b), "r"(c));
+    return res;
+}
+__device__ __forceinline__ int msub16(int a, int b, int c)
+{
+    int res;
+    asm("vmad.s32.u32.u32 %0, %1.h0, %2.h0, -%3;" : "=r"(res) : "r"(a), "r"(b), "r"(c));
+    return res;
+}
+)";
+    return ss.str();
+}
+
+std::string ngraph::runtime::gpu::nvrtc::define_zero(const std::string& dtype,
+                                                     const std::string& name)
+{
+    std::stringstream ss;
+    std::string zero = "static_cast<" + dtype + ">(0)";
+    ss << "__device__ __forceinline__ void " << name << "(" << dtype << "& a) { a = " << zero
+       << "; }\n";
+    return ss.str();
+}
+
+std::string ngraph::runtime::gpu::nvrtc::define_vzero(std::string dtype,
+                                                      const uint32_t& n,
+                                                      const std::string& name)
+{
+    std::stringstream ss;
+    if (n == 1 || n == 2 || n == 4)
+    {
+        std::string zero = "static_cast<" + dtype + ">(0)";
+        static std::vector<std::string> assignment = {"a.x = ", "a.y = ", "a.z = "};
+        dtype = dtype + std::to_string(n);
+        ss << "__device__ __forceinline__ void " << name << "(" << dtype << "& a) { ";
+        for (auto i = 0u; i <= (n >> 1); i++)
+        {
+            ss << assignment[i];
+        }
+        ss << zero << "; }\n";
+    }
+    else
+    {
+        throw ngraph_error("Invalid request for vector zero of " + dtype + std::to_string(n));
+    }
+    return ss.str();
+}
+#define LOAD_C                                                                                     \
+    "__device__ __forceinline__ " << dtype << " " << name << "(const " << dtype                    \
+                                  << "*  __restrict__ in, int i=0, bool b=true) { " << dtype       \
+                                  << "  v; zero_(v); if (b) v = in[i]; return v; }\n"
+
+std::string ngraph::runtime::gpu::nvrtc::define_coherent_load(const std::string& dtype,
```

Comment:
do you mean `aligned/unaligned` here with `coherent/noncoherent`?

Reply:
Nope, the `__ldg()` intrinsic is a read only load that is not coherent with global memory (updates are not propagated back to DRAM). Recently it has been implemented as a pull into constant memory for fast access that can help performance for memory bound kernels.

Reply:
`define_coherent_load` creates a load function that uses `ptr[offset]` for doing a coherent load from global memory.

`define_non_coherent_load` defines a load that uses `__ldg()`.

Reply:
We can change the name of these templates if that's preferred, I was just trying to be explicit. Maybe there is something better.

Reply:
ok, got, name is good if that's the case.